python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        num_freq = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_frame = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_meta = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_data = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        return (
+            num_writes,
+            num_time,
+            num_freq,
+            size_frame,
+            size_meta,
+            size_data,
+        )
+
+    def __del__(self):
+        self.semaphore.release()
+        try:
+            self.semaphore.unlink()
+        except posix_ipc.ExistentialError:
+            logger.debug("Semaphore file did not exist when trying to unlink from it.")
+
+        self.shared_mem.close()
+        try:
+            posix_ipc.unlink_shared_memory(self.shared_mem_name)
+        except posix_ipc.ExistentialError:
+            logger.debug(
+                "Shared memory file did not exist when trying to unlink from it."
+            )
+
+    def read_last(self, n: int):
+        """
+        Read last n time samples from the buffer.
+
+        Only copies from buffer if not read previously.
+
+        Parameters
+        ----------
+        n : int
+            Number of time samples to read.
+
+        Returns
+        -------
+        VisRaw
+            Last n time samples.
+
+        Raises
+        ------
+        SharedMemoryError
+            If the shared memory is marked as invalid by the writer or if the structural parameters
+            have changed.
+        """
+        if n > self.buffer_size:
+            raise SharedMemoryReaderUsageError(
+                "Can't read more than {0} last time slots, because buffer size is set to {0} "
+                "(Tried to read last {1} time slots.).".format(self.buffer_size, n)
+            )
+
+        self.shared_mem.seek(0)
+
+        self._validate_shm()
+
+        # get a data update from the ringbuffer
+        access_record = self._access_record()
+
+        times = self._filter_last(access_record, n)
+        logger.debug("Reading last {} time slots: {}".format(n, times))
+
+        self._copy_from_shm(times, access_record)
+
+        # check if any data became invalid while reading it
+        access_record_after_copy = self._access_record()
+        for t in times:
+            for f in range(self.num_freq):
+                if access_record_after_copy[t][f] != access_record[t][f]:
+                    logger.debug(
+                        "Data at t={}, f={} became invalid while reading it.".format(
+                            t, f
+                        )
+                    )
+                    self._data[self._time_index_map[t], f] = None
```

Comment:
This doesn't work, and is very inefficient. For one, you can't assign `None` to elements of `self._data`. I'm not sure how that passes your tests, so you might want to make sure your tests cover this.

This would be much better off done as something like:
```
self._data[...] = np.where(access_record_after_copy == access_record, 1, 0)
```

Now the `[...]` is because that doesn't work either with things as they currently are. I'll write the comment more above, but it probably makes a lot of sense to use a dtype that exposes the valid field, that way this could be `self._data["valid"] = ...`

Reply:
The chance of this happening in my tests is very low. I have to write a test for it still and then fix it.